devtools/x-lint/src/runner.rs

```diff
@@ -211,61 +238,12 @@ impl<'cfg> LintEngine<'cfg> {
-        // TODO: It'd be pretty cool to be able to cache the `&Path` instances as well. The options
-        // are:
-        // 1. Cache them as `PathBuf` instances. That would involve lots of allocations.
-        // 2. Store the `&Path` instances alongside `files_stdout`. That would be a self-referential
-        //    struct (so would require the use of something like
-        //    [`rental`](https://docs.rs/rental/).
-        // 3. Reorganize the code to use some sort of arena allocator which is passed into the lint
-        //    context (and/or stored alongside with `rental`.)
-        let stdout = self.files_stdout()?;
-
-        Ok(stdout
-            .split(|b| b == &0)
-            .filter_map(|s| {
-                // TODO: make global exclusions configurable.
-                if !s.is_empty() && !s.starts_with(b"testsuite/libra-fuzzer/artifacts/") {
-                    // `OsStr::from_bytes` only works on Unix, since "OS strings" on Windows are
-                    // actually UTF-16ish. Would be cool to get it working on Windows at some point
-                    // too, but it needs to be done with some care.
-                    //
-                    // For more, see:
-                    // * https://doc.rust-lang.org/std/ffi/index.html#conversions
-                    // * https://www.mercurial-scm.org/wiki/EncodingStrategy and
-                    // * https://en.wikipedia.org/wiki/Mojibake.
-                    use std::os::unix::ffi::OsStrExt;
-
-                    let s = OsStr::from_bytes(s);
-                    Some(Path::new(s))
-                } else {
-                    None
-                }
-            })
-            .collect())
-    }
-
-    fn files_stdout(&self) -> Result<&[u8]> {
-        self.files_stdout
-            .get_or_try_init(|| {
-                // TODO: abstract out SCM and command-running functionality.
-                let output = Command::new("git")
-                    .current_dir(self.config.core.project_root())
-                    // The -z causes files to not be quoted, and to be separated by \0.
-                    .args(&["ls-files", "-z"])
-                    .stderr(Stdio::inherit())
-                    .output()
-                    .map_err(|err| SystemError::io("running git ls-files", err))?;
-                if !output.status.success() {
-                    return Err(SystemError::Exec {
-                        cmd: "git ls-files",
-                        status: output.status,
-                    });
-                }
-                Ok(output.stdout)
-            })
-            .map(|stdout| stdout.as_slice())
+    fn file_list(&self) -> Result<impl Iterator<Item = &'cfg Path> + 'cfg> {
+        let tracked_files = self.config.core.git_cli().tracked_files()?;
+        // TODO: make global exclusions configurable
+        Ok(tracked_files
+            .iter()
+            .filter(|f| !f.starts_with("testsuite/libra-fuzzer/artifacts/")))
```

Comment:
Ah this is still needed do to header/final line rules and such.   Good TODO on it.